File: klib/source/math/ray/ray.cpp
#include "math/ray/ray.h"

#include "math/math.h"


// Construct
kl::ray::ray()
{}

kl::ray::ray(const float3& origin, const float3& direction)
    : origin(origin)
{
    set_direction(direction);
}

kl::ray::ray(const float3& origin, const float4x4& inverse_matrix, const float2& ndc)
    : origin(origin)
{
    float4 pixel_direction = inverse_matrix * float4(ndc, 1.0f, 1.0f);
    pixel_direction *= 1.0f / pixel_direction.w;
    set_direction(pixel_direction.xyz());
}

// Direction
void kl::ray::set_direction(const float3& direction)
{
    direction_ = normalize(direction);
}

kl::float3 kl::ray::get_direction() const
{
    return direction_;
}

// Intersection
bool kl::ray::intersect_plane(const plane& plane, float3* out_intersection) const
{
    const float denom = dot(plane.get_normal(), direction_);
    if (::abs(denom) <= 0.0001f) {
        return false;
    }

    const float t = dot(plane.origin - origin, plane.get_normal()) / denom;
    if (t < 0.0f) {
        return false;
    }

    if (out_intersection) {
        *out_intersection = origin + direction_ * t;
    }
    return true;
}

bool kl::ray::intersect_triangle(const triangle& triangle, float3* out_intersection) const
{
    const float3 edge1 = triangle.b.world - triangle.a.world;
    const float3 edge2 = triangle.c.world - triangle.a.world;

    const float3 h = cross(direction_, edge2);
    const float3 s = origin - triangle.a.world;
    const float f = 1.0f / dot(edge1, h);
    const float u = dot(s, h) * f;
    if (u < 0.0f || u > 1.0f) {
        return false;
    }

    const float3 q = cross(s, edge1);
    const float v = dot(direction_, q) * f;
    if (v < 0.0f || (u + v) > 1.0f) {
        return false;
    }

    const float t = dot(edge2, q) * f;
    if (t <= 0.0f) {
        return false;
    }

    if (out_intersection) {
        *out_intersection = origin + direction_ * t;
    }
    return true;
}

bool kl::ray::can_intersect_sphere(const sphere& sphere) const
{
    const float ray_distance = dot((sphere.origin - origin), direction_);
    const float3 ray_point = origin + direction_ * ray_distance;
    const float sphere_ray_distance = (sphere.origin - ray_point).length();
    return !(sphere_ray_distance > sphere.radius);
}

bool kl::ray::intersect_sphere(const sphere& sphere, float3* out_intersection, float* out_distance) const
{
    const float3 center_ray = sphere.origin - origin;
    const float cd_dot = dot(center_ray, direction_);
    if (cd_dot < 0.0f) {
        return false;
    }

    const float cc_dot = dot(center_ray, center_ray) - cd_dot * cd_dot;
    const float rr = sphere.radius * sphere.radius;
    if (cc_dot > rr) {
        return false;
    }

    const float thc = sqrt(rr - cc_dot);
    const float dis0 = cd_dot - thc;
    const float dis1 = cd_dot + thc;

    if (out_intersection) {
        *out_intersection = origin + direction_ * ((dis0 < 0.0f) ? dis1 : dis0);
    }
    if (out_distance) {
        *out_distance = (dis0 < 0.0f) ? dis1 : dis0;
    }
    return true;
}

// Format
std::ostream& kl::operator<<(std::ostream& os, const kl::ray& obj)
{
    os << "{" << obj.origin << ", " << obj.get_direction() << "}";
    return os;
}

File: script_sdk/source/sdk/math/ray/Ray.kt
package sdk.math.ray

import sdk.math.cross
import sdk.math.dot
import sdk.math.normalize
import sdk.math.triangle.Triangle
import sdk.math.vector.Float3
import sdk.math.vector.Float4
import java.io.Serializable
import kotlin.math.abs
import kotlin.math.sqrt

class Ray : Serializable {
    var origin: Float3
    var direction: Float3 = Float3()
        set(value) {
            field = normalize(value)
        }

    // Constructor
    constructor() : this(Float3(), Float3.POSITIVE_Z)

    constructor(origin: Float3, direction: Float3) {
        this.origin = origin
        this.direction = direction
    }

    // Intersect
    fun intersectPlane(plane: Plane, outIntersection: Float3?): Boolean {
        val denom = dot(plane.normal, direction)
        if (abs(denom) <= 0.0001f) {
            return false
        }

        val t = dot(plane.origin - origin, plane.normal) / denom
        if (t < 0f) {
            return false
        }

        outIntersection?.let {
            val inter = origin + direction * t
            it.x = inter.x
            it.y = inter.y
            it.z = inter.z
        }
        return true
    }

    fun intersectTriangle(triangle: Triangle, outIntersection: Float3?): Boolean {
        val edge1 = triangle.b.world - triangle.a.world
        val edge2 = triangle.c.world - triangle.a.world

        val h = cross(direction, edge2)
        val s = origin - triangle.a.world
        val f = 1f / dot(edge1, h)
        val u = dot(s, h) * f
        if (u < 0f || u > 1f) {
            return false
        }

        val q = cross(s, edge1)
        val v = dot(direction, q) * f
        if (v < 0f || (u + v) > 1f) {
            return false
        }

        val t = dot(edge2, q) * f
        if (t <= 0f) {
            return false
        }

        outIntersection?.let {
            val inter = origin + direction * t
            it.x = inter.x
            it.y = inter.y
            it.z = inter.z
        }
        return true
    }

    fun canIntersectSphere(sphere: Sphere): Boolean {
        val rayDistance = dot(sphere.origin - origin, direction)
        val rayPoint = origin + direction * rayDistance
        val sphereRayDistance = (sphere.origin - rayPoint).length
        return !(sphereRayDistance > sphere.radius)
    }

    fun intersectSphere(sphere: Sphere, outIntersection: Float4?): Boolean {
        val centerRay = sphere.origin - origin
        val cdDot = dot(centerRay, direction)
        if (cdDot < 0f) {
            return false
        }

        val ccDot = dot(centerRay, centerRay) - cdDot * cdDot
        val rr = sphere.radius * sphere.radius
        if (ccDot > rr) {
            return false
        }

        val thc = sqrt(rr - ccDot)
        val dis0 = cdDot - thc
        val dis1 = cdDot + thc

        outIntersection?.let {
            it.w = if (dis0 < 0f) dis1 else dis0
            it.x = origin.x + direction.x * it.w
            it.y = origin.y + direction.y * it.w
            it.z = origin.z + direction.z * it.w
        }
        return true
    }

    // Override
    override fun toString(): String {
        return "Ray($origin, $direction)"
    }
}
